### PR TITLE
Fix changing tabs closing query editor

### PR DESCRIPTION
**Pull request by apedroferreira**
Closes https://github.com/mui/mui-toolpad/issues/1742

- Fix query editor closing when switching browser tabs
- Also fixes console error when opening the query editor due to `margin: auto` in the page node.

https://user-images.githubusercontent.com/10789765/227025076-6d0c7a40-0dc5-4489-b46f-17c759865706.mov

### CODE REVIEW

File: packages/toolpad-app/src/toolpad/AppEditor/PageEditor/QueryEditor/index.tsx [hunk 1]
```diff
@@ -130,8 +130,12 @@ export default function QueryEditor() {
   const { queries = [] } = appDom.getChildNodes(dom, page) ?? [];
 
   const handleEditStateDialogClose = React.useCallback(() => {
-    appStateApi.setView({ kind: 'page', nodeId: page.id });
-  }, [appStateApi, page.id]);
+    if (!dialogState?.isDraft) {
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/AppEditor/PageEditor/QueryEditor/index.tsx:133]: Prefer positive logic
```suggestion
    if (dialogState?.isDraft) {
      setDialogState(null);
      // ...
```

File: packages/toolpad-app/src/toolpad/AppEditor/PageEditor/QueryEditor/index.tsx [hunk 2]
```diff
@@ -184,15 +188,19 @@ export default function QueryEditor() {
...
   React.useEffect(() => {
-    setDialogState(() => {
+    setDialogState((previousState) => {
       if (currentView.kind === 'page' && currentView.view?.kind === 'query') {
         const node = appDom.getNode(dom, currentView.view?.nodeId, 'query');
         return { node, isDraft: false };
       }
 
-      return null;
+      if (!dialogState?.isDraft) {
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/AppEditor/PageEditor/QueryEditor/index.tsx:197]: Same,

```suggestion
      if (dialogState?.isDraft) {
        return previousState;
      }
      
      return null;
```

File: packages/toolpad-app/src/toolpad/AppEditor/PageEditor/QueryEditor/index.tsx [hunk 3]
```diff
@@ -184,15 +188,19 @@ export default function QueryEditor() {
...
+        return null;
+      }
+
+      return previousState;
     });
-  }, [dom, currentView]);
+  }, [dom, currentView, dialogState?.isDraft]);
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/AppEditor/PageEditor/QueryEditor/index.tsx:203]: Perhaps extract this in a variable and reuse?

```tsx
const isDraft = !!dialogState?.isDraft;
```

**Review by Janpot — APPROVED** [review]
Few stylistic nits. Should we add/amend a test for this?

**Comment by apedroferreira** [comment]
> Should we add/amend a test for this?

Not sure, this was quite a specific issue since it involves switching tabs. The `dom` object changes when we switch tabs and it's a dependency in the effect that closes the query editor, so it was running unintentionally.
I can add a test, I don't think it's the most obvious case to write a test for, but if we don't mind the extra test it also offers extra coverage. I might just try to include switching tabs in an existing test for the query editor not to add a whole new test too, I'll give it a try.

**Comment by apedroferreira** [comment]
> Few stylistic nits. Should we add/amend a test for this?

I've modified the `rest-basics` test since it already tested the query editor a bit.

**Review by Janpot — APPROVED** [review]

**Pull request merged**